File: codex-rs/core/src/state/turn.rs
//! Turn-scoped state and active turn metadata scaffolding.

use indexmap::IndexMap;
use std::collections::HashMap;
use std::sync::Arc;
use std::time::Duration;
use std::time::Instant;
use tokio::sync::Mutex;
use tokio::sync::Notify;
use tokio_util::sync::CancellationToken;
use tokio_util::task::AbortOnDropHandle;

use codex_protocol::dynamic_tools::DynamicToolResponse;
use codex_protocol::models::ResponseInputItem;
use codex_protocol::request_user_input::RequestUserInputResponse;
use tokio::sync::oneshot;

use crate::codex::TurnContext;
use crate::protocol::ReviewDecision;
use crate::protocol::TokenUsage;
use crate::tasks::SessionTask;

/// Metadata about the currently running turn.
pub(crate) struct ActiveTurn {
    pub(crate) tasks: IndexMap<String, RunningTask>,
    pub(crate) turn_state: Arc<Mutex<TurnState>>,
}

impl Default for ActiveTurn {
    fn default() -> Self {
        Self {
            tasks: IndexMap::new(),
            turn_state: Arc::new(Mutex::new(TurnState::default())),
        }
    }
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub(crate) enum TaskKind {
    Regular,
    Review,
    Compact,
}

pub(crate) struct RunningTask {
    pub(crate) done: Arc<Notify>,
    pub(crate) kind: TaskKind,
    pub(crate) task: Arc<dyn SessionTask>,
    pub(crate) cancellation_token: CancellationToken,
    pub(crate) handle: Arc<AbortOnDropHandle<()>>,
    pub(crate) turn_context: Arc<TurnContext>,
    // Timer recorded when the task drops to capture the full turn duration.
    pub(crate) _timer: Option<codex_otel::Timer>,
}

#[derive(Debug, Default)]
pub(crate) struct TurnTimingState {
    started_at: Option<Instant>,
    first_token_at: Option<Instant>,
    first_message_at: Option<Instant>,
}

impl TurnTimingState {
    pub(crate) fn mark_turn_started(&mut self, started_at: Instant) {
        self.started_at = Some(started_at);
        self.first_token_at = None;
        self.first_message_at = None;
    }

    pub(crate) fn record_turn_ttft(&mut self) -> Option<Duration> {
        if self.first_token_at.is_some() {
            return None;
        }
        let started_at = self.started_at?;
        let first_token_at = Instant::now();
        self.first_token_at = Some(first_token_at);
        Some(first_token_at.duration_since(started_at))
    }

    pub(crate) fn record_turn_ttfm(&mut self) -> Option<Duration> {
        if self.first_message_at.is_some() {
            return None;
        }
        let started_at = self.started_at?;
        let first_message_at = Instant::now();
        self.first_message_at = Some(first_message_at);
        Some(first_message_at.duration_since(started_at))
    }
}

impl ActiveTurn {
    pub(crate) fn add_task(&mut self, task: RunningTask) {
        let sub_id = task.turn_context.sub_id.clone();
        self.tasks.insert(sub_id, task);
    }

    pub(crate) fn remove_task(&mut self, sub_id: &str) -> bool {
        self.tasks.swap_remove(sub_id);
        self.tasks.is_empty()
    }

    pub(crate) fn drain_tasks(&mut self) -> Vec<RunningTask> {
        self.tasks.drain(..).map(|(_, task)| task).collect()
    }
}

/// Mutable state for a single turn.
#[derive(Default)]
pub(crate) struct TurnState {
    pending_approvals: HashMap<String, oneshot::Sender<ReviewDecision>>,
    pending_user_input: HashMap<String, oneshot::Sender<RequestUserInputResponse>>,
    pending_dynamic_tools: HashMap<String, oneshot::Sender<DynamicToolResponse>>,
    pending_input: Vec<ResponseInputItem>,
    pub(crate) tool_calls: u64,
    pub(crate) token_usage_at_turn_start: TokenUsage,
}

impl TurnState {
    pub(crate) fn insert_pending_approval(
        &mut self,
        key: String,
        tx: oneshot::Sender<ReviewDecision>,
    ) -> Option<oneshot::Sender<ReviewDecision>> {
        self.pending_approvals.insert(key, tx)
    }

    pub(crate) fn remove_pending_approval(
        &mut self,
        key: &str,
    ) -> Option<oneshot::Sender<ReviewDecision>> {
        self.pending_approvals.remove(key)
    }

    pub(crate) fn clear_pending(&mut self) {
        self.pending_approvals.clear();
        self.pending_user_input.clear();
        self.pending_dynamic_tools.clear();
        self.pending_input.clear();
    }

    pub(crate) fn insert_pending_user_input(
        &mut self,
        key: String,
        tx: oneshot::Sender<RequestUserInputResponse>,
    ) -> Option<oneshot::Sender<RequestUserInputResponse>> {
        self.pending_user_input.insert(key, tx)
    }

    pub(crate) fn remove_pending_user_input(
        &mut self,
        key: &str,
    ) -> Option<oneshot::Sender<RequestUserInputResponse>> {
        self.pending_user_input.remove(key)
    }

    pub(crate) fn insert_pending_dynamic_tool(
        &mut self,
        key: String,
        tx: oneshot::Sender<DynamicToolResponse>,
    ) -> Option<oneshot::Sender<DynamicToolResponse>> {
        self.pending_dynamic_tools.insert(key, tx)
    }

    pub(crate) fn remove_pending_dynamic_tool(
        &mut self,
        key: &str,
    ) -> Option<oneshot::Sender<DynamicToolResponse>> {
        self.pending_dynamic_tools.remove(key)
    }

    pub(crate) fn push_pending_input(&mut self, input: ResponseInputItem) {
        self.pending_input.push(input);
    }

    pub(crate) fn take_pending_input(&mut self) -> Vec<ResponseInputItem> {
        if self.pending_input.is_empty() {
            Vec::with_capacity(0)
        } else {
            let mut ret = Vec::new();
            std::mem::swap(&mut ret, &mut self.pending_input);
            ret
        }
    }

    pub(crate) fn has_pending_input(&self) -> bool {
        !self.pending_input.is_empty()
    }
}

impl ActiveTurn {
    /// Clear any pending approvals and input buffered for the current turn.
    pub(crate) async fn clear_pending(&self) {
        let mut ts = self.turn_state.lock().await;
        ts.clear_pending();
    }
}

#[cfg(test)]
mod tests {
    use super::TurnTimingState;
    use std::time::Instant;

    #[test]
    fn turn_timing_state_records_ttft_only_once_per_turn() {
        let mut state = TurnTimingState::default();
        assert_eq!(state.record_turn_ttft(), None);

        state.mark_turn_started(Instant::now());
        assert!(state.record_turn_ttft().is_some());
        assert!(state.first_token_at.is_some());
        assert_eq!(state.record_turn_ttft(), None);
    }

    #[test]
    fn turn_timing_state_records_ttfm_independently_of_ttft() {
        let mut state = TurnTimingState::default();
        state.mark_turn_started(Instant::now());

        assert!(state.record_turn_ttft().is_some());
        assert!(state.record_turn_ttfm().is_some());
        assert!(state.first_token_at.is_some());
        assert!(state.first_message_at.is_some());
        assert_eq!(state.record_turn_ttfm(), None);
    }
}

File: src/nativeMain/kotlin/ai/solace/coder/core/session/Turn.kt
// port-lint: source core/src/state/turn.rs
package ai.solace.coder.core.session

import ai.solace.coder.client.auth.AuthManager
import ai.solace.coder.protocol.ReviewDecision
import ai.solace.coder.utils.concurrent.CancellationToken
import ai.solace.coder.protocol.TurnAbortReason
import ai.solace.coder.protocol.ResponseInputItem
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

/**
 * Turn-scoped state and active turn metadata scaffolding.
 *
 * Ported from Rust codex-rs/core/src/state/turn.rs
 */

/**
 * Describes the type of work a task performs.
 */
enum class TaskKind {
    Regular,
    Review,
    Compact
}

/**
 * Metadata about the currently running turn.
 */
class ActiveTurn {
    private val tasks = linkedMapOf<String, RunningTask>()
    val turnState = TurnState()

    fun addTask(task: RunningTask) {
        val subId = task.turnContext.subId
        tasks[subId] = task
    }

    /**
     * Remove a task by sub_id. Returns true if no tasks remain.
     */
    fun removeTask(subId: String): Boolean {
        tasks.remove(subId)
        return tasks.isEmpty()
    }

    fun drainTasks(): List<RunningTask> {
        val result = tasks.values.toList()
        tasks.clear()
        return result
    }

    fun getTasks(): Map<String, RunningTask> = tasks.toMap()

    /**
     * Clear any pending approvals and input buffered for the current turn.
     */
    suspend fun clearPending() {
        turnState.clearPending()
    }
}

/**
 * A running task in the session.
 */
data class RunningTask(
    val done: CompletableDeferred<Unit>,
    val kind: TaskKind,
    val task: SessionTask,
    val cancellationToken: CancellationToken,
    val turnContext: TurnContext
)

/**
 * Mutable state for a single turn.
 *
 * Ported from Rust codex-rs/core/src/state/turn.rs
 */
class TurnState {
    private val mutex = Mutex()
    private val pendingApprovals = mutableMapOf<String, CompletableDeferred<ReviewDecision>>()
    private val pendingInput = mutableListOf<ResponseInputItem>()

    /**
     * Insert a pending approval request.
     * Returns the previous deferred if one existed for this key.
     */
    suspend fun insertPendingApproval(
        key: String,
        deferred: CompletableDeferred<ReviewDecision>
    ): CompletableDeferred<ReviewDecision>? {
        return mutex.withLock {
            pendingApprovals.put(key, deferred)
        }
    }

    /**
     * Remove and return a pending approval by key.
     */
    suspend fun removePendingApproval(key: String): CompletableDeferred<ReviewDecision>? {
        return mutex.withLock {
            pendingApprovals.remove(key)
        }
    }

    /**
     * Clear all pending approvals and input.
     */
    suspend fun clearPending() {
        mutex.withLock {
            // Complete any pending approvals with Denied
            for ((_, deferred) in pendingApprovals) {
                if (!deferred.isCompleted) {
                    deferred.complete(ReviewDecision.Denied)
                }
            }
            pendingApprovals.clear()
            pendingInput.clear()
        }
    }

    /**
     * Push pending input item for the current turn.
     */
    suspend fun pushPendingInput(input: ResponseInputItem) {
        mutex.withLock {
            pendingInput.add(input)
        }
    }

    /**
     * Take all pending input, leaving the list empty.
     */
    suspend fun takePendingInput(): List<ResponseInputItem> {
        return mutex.withLock {
            if (pendingInput.isEmpty()) {
                emptyList()
            } else {
                val result = pendingInput.toList()
                pendingInput.clear()
                result
            }
        }
    }

    /**
     * Check if there are pending approvals.
     */
    suspend fun hasPendingApprovals(): Boolean {
        return mutex.withLock {
            pendingApprovals.isNotEmpty()
        }
    }
}

/**
 * Async task that drives a Session turn.
 *
 * Implementations encapsulate a specific Codex workflow (regular chat,
 * reviews, ghost snapshots, etc.). Each task instance is owned by a
 * Session and executed on a background coroutine.
 *
 * Ported from Rust codex-rs/core/src/tasks/mod.rs
 */
interface SessionTask {
    /**
     * Describes the type of work the task performs.
     */
    fun kind(): TaskKind

    /**
     * Executes the task until completion or cancellation.
     *
     * Implementations typically stream protocol events using session and ctx,
     * returning an optional final agent message when finished. The provided
     * cancellationToken is cancelled when the session requests an abort;
     * implementers should watch for it and terminate quickly once it fires.
     *
     * @return Optional final agent message
     */
    suspend fun run(
        sessionContext: SessionTaskContext,
        turnContext: TurnContext,
        input: List<UserInput>,
        cancellationToken: CancellationToken
    ): String?

    /**
     * Gives the task a chance to perform cleanup after an abort.
     *
     * The default implementation is a no-op; override this if additional
     * teardown or notifications are required once abort_all_tasks cancels the task.
     */
    suspend fun abort(sessionContext: SessionTaskContext, turnContext: TurnContext) {
        // Default no-op
    }
}

/**
 * Thin wrapper that exposes the parts of Session task runners need.
 *
 * Ported from Rust codex-rs/core/src/tasks/mod.rs SessionTaskContext
 */
class SessionTaskContext(
    private val session: Session
) {
    fun getSession(): Session = session

    /**
     * Get the auth manager for API authentication.
     * Ported from Rust codex-rs/core/src/tasks/mod.rs SessionTaskContext::auth_manager
     */
    fun authManager(): AuthManager = session.services.authManager
}

/**
 * User input for a turn.
 *
 * Ported from Rust codex-rs/protocol/src/user_input.rs
 */
sealed class UserInput {
    /**
     * Text input from the user.
     */
    data class Text(val content: String) : UserInput()

    /**
     * Image input from the user.
     */
    data class Image(
        val data: ByteArray,
        val mimeType: String
    ) : UserInput() {
        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (other !is Image) return false
            return data.contentEquals(other.data) && mimeType == other.mimeType
        }

        override fun hashCode(): Int {
            var result = data.contentHashCode()
            result = 31 * result + mimeType.hashCode()
            return result
        }
    }

    /**
     * File reference input.
     */
    data class FileRef(val path: String) : UserInput()
}

// TurnAbortReason is imported from ai.solace.coder.protocol
